File: api-clients/elixir/lib/otreeba_open_cannabis_api/api/strains.ex
defmodule OtreebaOpenCannabisAPI.Api.Strains do
  @moduledoc """
  Documentation for OtreebaOpenCannabisAPI.Api.Strains.
  """

  use Tesla

  plug Tesla.Middleware.BaseUrl, "https://api.otreeba.com/v1"
  plug Tesla.Middleware.JSON

  def get_strain_by_ocpc(ocpc) do
    method = [method: :get]
    url = [url: "/strains/#{ocpc}"]
    query_params = []
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  def get_strains(page, count, sort) do
    method = [method: :get]
    url = [url: "/strains"]
    query_params = [query: [{:"page", page}, {:"count", count}, {:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end
end

File: api-clients/clojure/src/otreeba_open_cannabis_api/api/strains.clj
(ns otreeba-open-cannabis-api.api.strains
  (:require [otreeba-open-cannabis-api.core :refer [call-api check-required-params with-collection-format]])
  (:import (java.io File)))

(defn get-strain-by-ocpc-with-http-info
  "Find strain by Open Cannabis Product Code (OCPC).
  Returns a single strain."
  [ocpc ]
  (call-api "/strains/{ocpc}" :get
            {:path-params   {"ocpc" ocpc }
             :header-params {}
             :query-params  {}
             :form-params   {}
             :content-types ["application/json"]
             :accepts       ["application/json"]
             :auth-names    ["api_key"]}))

(defn get-strain-by-ocpc
  "Find strain by Open Cannabis Product Code (OCPC).
  Returns a single strain."
  [ocpc ]
  (:data (get-strain-by-ocpc-with-http-info ocpc)))

(defn get-strains-with-http-info
  "Get a list of all current strains.
  Returns a paginated list of strains."
  ([] (get-strains-with-http-info nil))
  ([{:keys [page count sort ]}]
   (call-api "/strains" :get
             {:path-params   {}
              :header-params {}
              :query-params  {"page" page "count" count "sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    ["api_key"]})))

(defn get-strains
  "Get a list of all current strains.
  Returns a paginated list of strains."
  ([] (get-strains nil))
  ([optional-params]
   (:data (get-strains-with-http-info optional-params))))

